src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.Tests/ServiceControllerTests.cs

```diff
@@ -0,0 +1,260 @@
+using System.ServiceProcess;
+using System.Threading;
+using Xunit;
+using Microsoft.Win32;
+
+namespace System.ServiceProcessServiceController.Tests
+{
+    internal sealed class ServiceProvider
+    {
+        public readonly string TestMachineName;
+        public readonly TimeSpan ControlTimeout;
+        public readonly string TestServiceName;
+        public readonly string TestServiceDisplayName;
+        public readonly string DependentTestServiceNamePrefix;
+        public readonly string DependentTestServiceDisplayNamePrefix;
+        public readonly string TestServiceRegistryKey;
+
+        public ServiceProvider()
+        {
+            TestMachineName = ".";
+            ControlTimeout = TimeSpan.FromSeconds(3);
+            TestServiceName = Guid.NewGuid().ToString();
+            TestServiceDisplayName = "Test Service " + TestServiceName;
+            DependentTestServiceNamePrefix = TestServiceName + ".Dependent";
+            DependentTestServiceDisplayNamePrefix = TestServiceDisplayName + ".Dependent";
+            TestServiceRegistryKey = @"HKEY_USERS\.DEFAULT\ToF\ServiceController\" + TestServiceName;
+
+            // Create the service
+            CreateTestServices();
+        }
+
+        private void CreateTestServices()
+        {
+            // Create the test service and its dependent services. Then, start the test service.
+            // All control tests assume that the test service is running when they are executed.
+            // So all tests should make sure to restart the service if they stop, pause, or shut
+            // it down.
+            RunServiceExecutable("create");
+        }
+
+        public void DeleteTestServices()
+        {
+            RunServiceExecutable("delete");
+            RegistryKey users = Registry.Users;
+            if (users.OpenSubKey(".DEFAULT\\ToF") != null)
+                users.DeleteSubKeyTree(".DEFAULT\\ToF");
+        }
+
+        private void RunServiceExecutable(string action)
+        {
+            var process = new Process();
+            process.StartInfo.FileName = "NativeTestService.exe";
+            process.StartInfo.Arguments = string.Format("\"{0}\" \"{1}\" {2}", TestServiceName, TestServiceDisplayName, action);
+            process.Start();
+            process.WaitForExit();
+            
+            if (process.ExitCode != 0)
+            {
+                throw new Exception("error: NativeTestService.exe failed with exit code " + process.ExitCode.ToString());
+            }
+        }
+    }
+
+    public class ServiceControllerTests : IDisposable
+    {
+        ServiceProvider _testService; 
+
+        public ServiceControllerTests()
+        {
+            _testService = new ServiceProvider();
+        }
+
+        [Fact]
+        public void ConstructWithServiceName()
+        {
+            var controller = new ServiceController(_testService.TestServiceName);
+            Assert.Equal(_testService.TestServiceName, controller.ServiceName);
```

Comment:
In all of these tests, should we be unconditionally shutting down the service in a finally?  I worry we're going to have runaway services in our testing infrastructure, similar to issues we've faced with System.Diagnostics.Process.Tests.